libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,211 @@
+#include "soft_timer.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "critical_section.h"
+#include "interrupt.h"
+#include "status.h"
+#include "stm32f0xx_interrupt.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_tim.h"
+
+typedef struct Timer {
+  uint32_t duration;
+  void *context;
+  bool inuse;
+  SoftTimerCallback callback;
+} Timer;
+
+static volatile Timer s_soft_timer_array[SOFT_TIMER_MAX_TIMERS];
+static volatile SoftTimerID s_active_timer_id;
+static volatile uint32_t s_last_update_count;
+static volatile uint32_t s_active_timer_duration;
+
+static void prv_start_timer(uint32_t duration_us) {
+  // Clear any pending interrupts, this will always occur in an interrupt or new timer getting
+  // started in a critical section so we need to prevent a accidental trigger if the time expires in
+  // the middle of the critical section and we started a new timer which falls before it.
+  TIM_ClearITPendingBit(TIM2, TIM_IT_CC1);
+
+  TIM_Cmd(TIM2, DISABLE);
```

Comment:
The reason I suggested using a separate hardware timer is so it can always run. If you're disabling and reenabling the timer to modify its counter, you lose some precision.